[#450] friend vs close friend implementation for ver.D and ver.E Common Features

tyzhou05 opened the pull request:
## Issue Number:

## Self Check List

- [x] Have you double-checked the base branch into which this PR will be merged?
- [x] Have you rebased from the base branch?

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (formatting, renaming)
- [ ] Merge Develop to Main
- [ ] Other (please describe):

## What does this PR do?

- Added visibility field in Note and Response models, defaults to friends but allows for close_friends (except for ver.D Note)
- Note model now should have two versions. ver.D: Always set to friends visibility and ver.E: users can now choose visibility(regular endpoints)
- updated ver.D Note endpoints as /api/notes/default/ and /api/notes/<int:pk>/default/
- Response model now has only Version E with user choice of visibility

## Preview Image

## Further comments

tyzhou05 committed:
initial commit

njs03332 commented on adoorback/qna/models.py:
```diff
@@ -76,6 +76,11 @@ class Meta:
+    visibility = models.CharField(
+        max_length=20,
+        choices=[('friends', 'Friends'), ('close_friends', 'Close Friends')],
+        default='friends'
```
For ver.E, the default visibility should be set to **close friends**!
While this will be reflected in the frontend UI as the initially selected default value ([Reference](https://www.notion.so/jaewon-kim/Connection-161f7ff219e581fd944ac0f746877ef5?pvs=4#20081375b1a64484bf4c55382e3f550a)), I think it makes more sense to unify this default as `close_friends` in the backend as well.
Was there any specific reason for setting the default to friends?

tyzhou05 replied:
Ahh I see, I think I misunderstood the spec - I agree it makes more sense to unify with the frontend's default of `close_friends` for ver.E! Made the change

njs03332 commented on adoorback/qna/serializers.py:
```diff
@@ -43,6 +43,7 @@ class ResponseSerializer(AdoorBaseSerializer):
     current_user_read = serializers.SerializerMethodField(read_only=True)
     current_user_reaction_id_list = serializers.SerializerMethodField(read_only=True)
     like_reaction_user_sample = serializers.SerializerMethodField(read_only=True)
+    visibility = serializers.ChoiceField(choices=['friends', 'close_friends'], default='friends')
```
I think it would be safer to enforce the frontend to always send the `visibility` value by removing `default='friends'` here and setting `required=True`!

njs03332 commented on adoorback/note/serializers.py:
```diff
@@ -22,6 +22,7 @@ class NoteSerializer(AdoorBaseSerializer):
     current_user_read = serializers.SerializerMethodField(read_only=True)
     current_user_reaction_id_list = serializers.SerializerMethodField(read_only=True)
     like_reaction_user_sample = serializers.SerializerMethodField(read_only=True)
+    visibility = serializers.ChoiceField(choices=['friends', 'close_friends'], default='friends')
```
Same here as well!

njs03332 commented on adoorback/note/serializers.py:
```diff
@@ -105,6 +106,14 @@ def get_like_user_sample(self, obj):
         return obj.liked_user_ids.count()
 
+    def create(self, validated_data):
+        validated_data['visibility'] = 'friends'
```
This shouldn't be forced to 'friends'.  

To clarify, the term **"Default post visibility"** doesn't mean that the value is always set to the default, but rather that it is the value applied if the user doesn't explicitly change it on the visibility selection screen.  
In other words, if a user selects **'close friends'** instead of the default **'friends'** in ver.D, the frontend should send `'close_friends'` as the visibility value, and the system should set the visibility accordingly.  

Let me know if I’ve misunderstood anything!

tyzhou05 replied:
Ahh I see, thanks! I think I originally misunderstood the "Default post visibility" requirement. The frontend should handle showing '`friends`' as the default option in ver.D, but users should still be able to change it to '`close_friends`' if they want to, correct?

I'll remove the forced 'friends' visibility from the backend, and I think my changes allow it so that the backend will accept whichever visibility value the frontend sends!

njs03332 replied:
I think here in the `create` and update `methods`, you're still forcing `visibility='friends'`! 
This should be removed.

njs03332 commented on adoorback/note/views.py:
```diff
@@ -225,3 +225,16 @@ def patch(self, request, *args, **kwargs):
         return Response(serializer.data)
+
+
+class DefaultFriendNoteCreate(NoteCreate):
```
As mentioned above, since visibility doesn’t need to be forced to 'friends' in the backend, I don't think we need a separate API for `NoteCreate` !

njs03332 reviewed (changes requested):
Thanks for the quick work! :) 

Just one more comment: It would be nice to add a constraint to check whether the **current user** has access to the post using the `is_audience` function and throw an error if they don’t.  
Currently, this is implemented in views like **ResponseComments** and **NoteComments**, but it would be helpful to add it to the following views as needed:  

- **NoteDetail**  
- **DefaultFriendNoteDetail**  
- **NoteLikes**  
- **ResponseDetail**  

This ensures that even if a user without access tries to access the URL, they won’t be able to see the post.

tyzhou05 committed:
updating visibilities; adding new constraints

tyzhou05 commented:
Overall changes:
* Set default visibility to `close_friends` in both Note and Response models
* Added `required=True` to visibility fields
* Removed default values to enforce frontend sending visibility
* Removed forced '`friends`' visibility from `DefaultFriendNoteSerializer`


Simplified API endpoints:
* Removed separate /default/ endpoints
* Removed unnecessary `DefaultFriendNoteCreate` view

Added new constraints:
* Raises PermissionDenied error if the **current user** doesn't have access using `is_audience`

And if I understand correctly, frontend will handle default visibility selection. Let me know if the changes look good or not!

njs03332 commented on adoorback/note/urls.py:
```diff
@@ -5,7 +5,6 @@
 urlpatterns = [
     path('', views.NoteCreate.as_view(), name='note-list'),
     path('<int:pk>/', views.NoteDetail.as_view(), name='note-detail'),
-    path('<int:pk>/default/', views.DefaultFriendNoteDetail.as_view(), name='default-friend-note-detail'),
```
This API should not be removed, it was implemented [here](https://github.com/ssm0318/WhoamI-Today-backend/pull/429) because we needed a separate API for friend note detail in ver.D! (Due to difference in fields `like_user_sample` / `like_reaction_user_sample` in the two versions.)

tyzhou05 replied:
Ah yes that was my bad, done!

njs03332 commented on adoorback/note/views.py:
```diff
@@ -102,6 +102,9 @@ def get_object(self):
 
         self.check_object_permissions(self.request, obj)
 
+        if not obj.is_audience(self.request.user):
```
This should also be added in `DefaultFriendNoteDetail` API!

njs03332 commented on adoorback/qna/migrations/0008_response_visibility.py:
```diff
@@ -0,0 +1,18 @@
+        migrations.AddField(
+            model_name='response',
+            name='visibility',
+            field=models.CharField(choices=[('friends', 'Friends'), ('close_friends', 'Close Friends')], default='friends', max_length=20),
```
The default here should be changed to `'close_friends'` as well! 

You can either delete this migration file and run `python manage.py makemigrations` again to recreate a migration file based on your changed code,
or just manually change the default in this migration file to `'close_friends'` if that's the only change you made to the model after running `makemigrations`.

njs03332 reviewed (commented):
It looks like you merged it while I was still leaving comments on the code! 😅 (I think the emojis I left on your responses might have been misleading—sorry if that caused any confusion!)

For next time, it might be helpful to keep these two principles in mind:
1) The rule is to merge into `main` only after at least one reviewer has **approved**. (Right now, since our backend team is small, I've been merging without reviews. But in your case, it's best to merge only after I approve!)
2) Merging into `main` should only happen once we've confirmed that everything works correctly. (I'm not entirely sure—have all the APIs implemented here been tested in Postman? If they have, then you've passed this step!)

I just reverted the merge to main [here](https://github.com/ssm0318/WhoamI-Today-backend/pull/455) and restored the `common2` branch!

It'd be great if you could create a new PR with `common2` branch, review the comments I left, apply any necessary changes, and then request me for review again! Also, once the implementation is complete, before merging, it’s important to summarize the changes in the top comment of the PR so the frontend team can easily review them!  ([Reference](https://github.com/ssm0318/WhoamI-Today-backend/pull/445#issue-2810927354))

Let me know if I misunderstood anything! 😊
